Mathlib/Order/OrdinalApproximantsFixedPoints.lean

```diff
@@ -0,0 +1,365 @@
+  · exact h_fab
+
+/-- If there are ordinals with equal value then the values are fixed points of f -/
+lemma lfp_approx_one_fixedPoint (a b : Ordinal.{u}) (h : a < b)
+    (h_fab : lfp_approx f a = lfp_approx f b):
+    lfp_approx f a ∈ (fixedPoints f) := by
+  rw[mem_fixedPoints_iff,lfp_approx_addition]
+  apply Monotone.monotone_stabilizing (lfp_approx f) (lfp_approx_monotone f)
+    h_fab (a+1) (SuccOrder.succ_le_of_lt h) (SuccOrder.le_succ a)
+
+/-- If there are ordinals with equal value then
+  every value follwing the smaller ordinal are fixed points of -/
+lemma lfp_approx_many_fixedPoints (a b : Ordinal.{u}) (h : a < b)
+    (h_fab : lfp_approx f a = lfp_approx f b):
+    ∀ i ≥ a, lfp_approx f i ∈ (fixedPoints f) := by
+  intro i h_i;
+  by_cases h_ia : i = a
+  · rw[h_ia]; exact lfp_approx_one_fixedPoint f a b h h_fab
+  · apply lfp_approx_eq_fp f a
+    · exact lfp_approx_one_fixedPoint f a b h h_fab
+    · exact Ne.lt_of_le' h_ia h_i
+
+/-- A fixed point of f is reached after the successor of the domains cardinality -/
+theorem lfp_approx_has_fixedPoint_cardinal : lfp_approx f (ord $ succ #α) ∈ (fixedPoints f) := by
+  let ⟨a, h_a, b, h_b, h_nab, h_fab⟩ := lfp_approx_equal_value f
+  cases (le_total a b) with
+  | inl h_ab =>
+    exact lfp_approx_many_fixedPoints f a b (Ne.lt_of_le h_nab h_ab) h_fab
+      (ord $ succ #α) (le_of_lt h_a)
+  | inr h_ba =>
+    exact lfp_approx_many_fixedPoints f b a (Ne.lt_of_le (id (Ne.symm h_nab)) h_ba)
+      (h_fab.symm) (ord $ succ #α) (le_of_lt h_b)
+
+/-- Every value of the ordinal approximants are less or equal than every fixed point of f -/
+theorem lfp_approx_le_fixedPoint : ∀ a : (fixedPoints f), ∀ i : Ordinal, lfp_approx f i ≤ a := by
+  intro ⟨a, h_a⟩ i
+  induction i using Ordinal.induction with
+  | h i IH =>
+    unfold lfp_approx
+    apply sSup_le; simp
+    intro j h_j
+    rw[← h_a]
+    apply f.monotone'
+    exact IH j h_j
+
+/-- The least fixed point of f is reached after the successor of the domains cardinality -/
+theorem lfp_approx_cardinal_is_lfp : lfp_approx f (ord $ succ #α) = lfp f := by
+  apply le_antisymm
+  · have h_lfp : ∃ x : fixedPoints f, lfp f = x := by use ⊥; exact rfl
+    let ⟨x, h_x⟩ := h_lfp; rw[h_x]
+    exact lfp_approx_le_fixedPoint f x (ord $ succ #α)
+  · have h_fix : ∃ x: fixedPoints f, lfp_approx f (ord $ succ #α) = x := by
+      simpa using lfp_approx_has_fixedPoint_cardinal f
+    let ⟨x, h_x⟩ := h_fix; rw[h_x]
+    exact lfp_le_fixed f x.prop
+
+/-- **Constructive Knaster-Tarski Theorem**: Some ordinal approximation of the least fixed point
+  is the least fixed point. Also known as ordinal approximation of the least fixed point.-/
+theorem lfp_approx_is_lfp : ∃ a : Ordinal, lfp_approx f a = lfp f := by
+  use (ord $ succ #α)
+  exact lfp_approx_cardinal_is_lfp f
+
+
+set_option linter.unusedVariables false in
+/-- Ordinal approximants of the least fixed points -/
+def gfp_approx (a : Ordinal.{u}) : α :=
```

Comment:
Seems like a lot of repeated code after this. Maybe there's a clever way to avoid it?

Reply:
I use now the order homomorphism duality.